File: 2020/02/02.go
package main

import "io/ioutil"
import "fmt"
import "strings"

func parseLine(line string) (int, int, string, string) {
	var a, b int
	var letter, password string
	fmt.Sscanf(line, "%d-%d %1s: %s", &a, &b, &letter, &password)
	return a, b, letter, password
}

func main() {
	data, _ := ioutil.ReadFile("input.txt")

	one := 0
	two := 0
	for _, s := range strings.Split(string(data), "\n") {
		a, b, letter, password := parseLine(s)
		if len(password) == 0 {
			continue
		}

		cnt := strings.Count(password, letter)
		if a <= cnt && cnt <= b {
			one++
		}

		aa := string(password[a-1])
		bb := string(password[b-1])
		if (aa == letter) != (bb == letter) {
			two++
		}
	}
	fmt.Println("Part 1:", one)
	fmt.Println("Part 2:", two)
}

File: 2020/02/02.py
import re

PATTERN = re.compile(r'(\d+)-(\d+) ([a-z]): ([a-z]+)')

def parse(line):
    a, b, letter, pw = re.match(PATTERN, line).groups()
    return int(a), int(b), letter, pw

with open('input.txt') as f:
    passwords = [parse(line) for line in f.readlines()]

one = 0
two = 0
for a, b, letter, pw in passwords:
    if a <= pw.count(letter) <= b:
        one += 1
    if (pw[a - 1] == letter) ^ (pw[b - 1] == letter):
        two += 1

print('Part 1:', one)
print('Part 2:', two)
